compiler/rustc_borrowck/src/borrowck_errors.rs

```diff
@@ -351,53 +347,38 @@ impl<'cx, 'tcx> crate::MirBorrowckCtxt<'cx, 'tcx> {
         immutable_section: &str,
         action: &str,
     ) -> DiagnosticBuilder<'cx, ErrorGuaranteed> {
-        let mut err = struct_span_err!(
-            self,
-            mutate_span,
-            E0510,
-            "cannot {} {} in {}",
-            action,
-            immutable_place,
-            immutable_section,
-        );
-        err.span_label(mutate_span, format!("cannot {}", action));
-        err.span_label(immutable_span, format!("value is immutable in {}", immutable_section));
-        err
+        let err = MutateInImmute { action, immutable_place, immutable_section, mutate_span };
+        let mut diag = self.infcx.tcx.sess.create_err(err);
+        diag.subdiagnostic(ImmutableValueLabel::Here { immutable_span });
+        diag
     }
 
     pub(crate) fn cannot_borrow_across_generator_yield(
         &self,
         span: Span,
         yield_span: Span,
     ) -> DiagnosticBuilder<'cx, ErrorGuaranteed> {
-        let mut err = struct_span_err!(
-            self,
-            span,
-            E0626,
-            "borrow may still be in use when generator yields",
-        );
-        err.span_label(yield_span, "possible yield occurs here");
-        err
+        let err = BorrowAcrossGeneratorYield { span };
+        let mut diag = self.infcx.tcx.sess.create_err(err);
+        diag.subdiagnostic(PossibleYieldLabel::Here { yield_span });
```

Comment:
As before, `BorrowAcrossGeneratorYield` might be able to use `#[label]` directly or at least have the subdiagnostic as a field.